Buffer refactor

approxit opened the pull request:
What I've done:
* Refactored whole buffer and scoring buffer from `golem.managers.proposals.plugins`
* Introduced composable buffers - simple buffer / background fill / expiration
* Introduced `Waiter` to take the heavy lifting from simple buffer
* Introduced `SingleUseSemaphore` to take the heavy lifting from bacground fill buffer
* Moved asyncio related stuff from `golem.utils` to single name space `golem.utils.asyncio`, as stuff started to be mixed with other utils
* Tests for coverage
* Reformatted the code

Notable remarks:
* Low level asyncio is tricky, but quite fun to play around :D
* `SimpleBuffer` covers quite fully functionality from `ErrorReportingQueue`, I would remove the second one in favor of the first, but team have different opinions about this, so it will be left as is
* PR will be changed from draft after ray-on-golem testing session

lucekdudek commented on golem/managers/proposal/plugins/scoring/scoring_buffer.py:
```diff
@@ -1,33 +1,26 @@
 
-from golem.managers.proposal.plugins.buffer import Buffer
-from golem.managers.proposal.plugins.scoring.mixins import ProposalScoringMixin
+from golem.managers.proposal.plugins.buffer import Buffer as BufferPlugin
```
Maybe we should rename `class Buffer(ProposalManagerPlugin):` to `BufferPlugin` now that we have `BackgroundFillBuffer` and others?

approxit commented:
In addition to main PR descriptions, here are more stuff done:

* Fixed some log lines to use `%` instead of f-strings.
* Use `cancel_and_await`/`cancel_and_await_many` when applicable.
* Renamed `Buffer` plugin to `BufferPlugin` from managers level
* Fixed checks, which require some lib upgrades, and also fix version of python
* Added python 3.12 in CI
* Fix python 3.12 checks complaints

shadeofblue commented on golem/managers/proposal/plugins/buffer.py:
```diff
@@ -1,130 +1,103 @@
 
 
-class Buffer(ProposalManagerPlugin):
+class BufferPlugin(ProposalManagerPlugin):
```
shouldn't the name here be `ProposalBufferPlugin`? what if we have a `BufferPlugin` in some other manager?

approxit replied:
You touched quite interesting topic that have some pros and cons - I'd like to talk it through on some meeting.

approxit commented:
And another update:
* `Demand` and `Proposal` resources got new method to get their expiration dates
* `RefreshingDemandManager` uses deamnd's expiration date instead of misused property
* `Buffer` got remixed interface to support race conditions in buffer composition.

shadeofblue commented on golem/utils/asyncio/buffer.py:
```diff
@@ -0,0 +1,403 @@
+        else:
+            yield
+
+    @trace_span()
```
I don't really feel why we need to have those decorators on such low level objects... it's as if we wanted to have such logging on e.g. asyncio.Queue objects... 

plus, they're used inconsistenly, as far as I see...

shadeofblue commented on golem/utils/typing.py:
```diff
@@ -1,4 +1,8 @@
+
+T = TypeVar("T")
+
+MaybeAwaitable = Union[Awaitable[T], T]
```
👍

shadeofblue reviewed (commented):
whew... that's a lot of code ... I don't feel I have finished the review and probably it would be good to touch base on a call anyway but here are some comments that I have had...

approxit commented on golem/utils/asyncio/queue.py:
```diff
@@ -1,21 +1,21 @@
-class ErrorReportingQueue(asyncio.Queue, Generic[QueueItem]):
-    """Asyncio Queue that enables exceptions to be passed to consumers from the feeding code."""
 
-    _error: Optional[BaseException]
```
Change is based on existing patterns found in the repository - keeping typings inline for fields defines in `__init__`

lucekdudek commented on golem/managers/proposal/plugins/buffer.py:
```diff
@@ -1,130 +1,103 @@
         min_size: int,
         max_size: int,
-        fill_concurrency_size: int = 1,
+        fill_concurrency_size=1,
```
@approxit ?

shadeofblue commented on golem/managers/proposal/plugins/buffer.py:
```diff
@@ -99,32 +100,31 @@ async def _request_items(self) -> None:
-                "Target items is now `%s` which is not below min size `%d`, requesting fill not"
-                " needed",
-                items_count,
+                "Proposals count %d is not below minimum size %d, skipping fill",
```
why "not below" instead of "above"?

```suggestion
                "Proposals count %d is above minimum size %d, skipping fill",
```

shadeofblue commented on golem/managers/proposal/plugins/buffer.py:
```diff
@@ -99,32 +100,31 @@ async def _request_items(self) -> None:
         return proposal
 
-    async def _get_item(self) -> Proposal:
+    async def _get_buffered_proposal(self) -> Proposal:
```
why do we need the adjective `buffered` ? is there a way on this class to get a non-buffered proposal? 

```suggestion
    async def _get_proposal(self) -> Proposal:
```

approxit replied:
`get_proposal` and `_get_proposal` are already in use by base `ProposalManagerPlugin` class, so we can't use them here. We need different names.

shadeofblue commented on golem/managers/proposal/plugins/buffer.py:
```diff
@@ -99,32 +100,31 @@ async def _request_items(self) -> None:
         return await self._buffer.get()
 
-    def _get_items_count(self) -> int:
+    def _get_buffered_proposals_count(self) -> int:
```
same - is there a way to get a non-buffered proposals count?

```suggestion
    def _get_proposals_count(self) -> int:
```

shadeofblue commented on golem/managers/proposal/plugins/scoring/scoring_buffer.py:
```diff
@@ -55,33 +75,34 @@ async def stop(self) -> None:
-            items = await self._buffer.get_all_requested(self._update_interval)
-            logger.debug("Waiting for more items done, %d new items will be scored", len(items))
+            logger.debug(
+                "Waiting for any proposals to score with debounce of `%s`...",
```
I'm unsure if "debounce" is the most obvious word in this context... I'm not even sure what it means here... is this the minimum number of proposals to wait for until any scoring will happen?

approxit replied:
`Debounce` is fairly settled term, it always refers to time. Minimum amount of time to catch more proposals before scoring will happen.

shadeofblue replied:
far from perfectly clear but indeed, I hadn't looked at the type there

shadeofblue reviewed (approved):
please see my comments... otherwise, lgtm afaict
